### PR TITLE
MaintainIPPool refactor and support for aborting release handshake

### PR DESCRIPTION
**Pull request by hemanthmalla**
Imagine a scenario where a node has 2 unused IPs and pre-allocate set to one. Let's say one of the IPs is in the middle of a handshake and a new pod is scheduled on the node. The other unused IP would be allocated to the pod. Now, when the operator re-evaluates, the node is no longer considered to be in excess. Without this commit, the operator does not act further on IPs in this state. This results in a scenario where no new IPs are allocated to the node and agent cannot allocate the unused IPs because they're in the middle of a handshake.

Also refactors IP release and allocate functionality out of maintainIPPool()

```release-note
Fix for a bug where unused IPs on the node cannot be allocated when IP release handshake is enabled. Adds support for aborting IP release, if the node doesn't have excess anymore.
```

### CODE REVIEW

File: pkg/ipam/node.go [hunk 1]
```diff
@@ -586,6 +586,30 @@ func (n *Node) removeStaleReleaseIPs() {
 	}
 }
 
+// abortNoLongerExcessIPs allows for aborting release of IP if new allocations on the node result in a change of excess
+// count or the interface selected for release.
+func (n *Node) abortNoLongerExcessIPs(excessMap *map[string]bool) {
```
Comment (gandro) [inline pkg/ipam/node.go:591]:
```suggestion
func (n *Node) abortNoLongerExcessIPs(excessMap map[string]bool) {
```

Comment (gandro) [inline pkg/ipam/node.go]:
```suggestion
func (n *Node) abortNoLongerExcessIPs(excessMap map[string]bool) {
```

File: pkg/ipam/node.go [hunk 2]
```diff
@@ -586,6 +586,30 @@ func (n *Node) removeStaleReleaseIPs() {
...
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	if n.resource.Status.IPAM.ReleaseIPs == nil || len(n.resource.Status.IPAM.ReleaseIPs) == 0 {
```
Comment (gandro) [inline pkg/ipam/node.go]: No need for a nil check here, `len(nil)` is safe and will always return 0.
```suggestion
	if len(n.resource.Status.IPAM.ReleaseIPs) == 0 {
```

File: pkg/ipam/node.go [hunk 3]
```diff
@@ -586,6 +586,30 @@ func (n *Node) removeStaleReleaseIPs() {
...
+		return
+	}
+	for ip, status := range n.resource.Status.IPAM.ReleaseIPs {
+		if excessMap != nil && (*excessMap)[ip] {
```
Comment (gandro) [inline pkg/ipam/node.go]:
```suggestion
		if excessMap[ip] {
```

Read-access to a nil map is safe and will work like the empty map.

File: pkg/ipam/node.go [hunk 4]
```diff
@@ -682,14 +704,25 @@ func (n *Node) maintainIPPool(ctx context.Context) (instanceMutated bool, err er
...
+	// Abort handshake for IPs that are in the middle of handshake, but are no longer considered excess
+	if a.release != nil && len(a.release.IPsToRelease) > 0 {
+		excessMap := make(map[string]bool)
+		for _, ip := range a.release.IPsToRelease {
+			excessMap[ip] = true
+		}
+		n.abortNoLongerExcessIPs(&excessMap)
+	} else {
+		n.abortNoLongerExcessIPs(nil)
+	}
```
Comment (gandro) [inline pkg/ipam/node.go]: tiny nit (feel free to ignore): I prefer this form without an else-statement (requires the abortNoLongerExcessIPs signature change mentioned above)

```go
var excessMap map[string]bool
if a.release != nil && len(a.release.IPsToRelease) > 0 {
	excessMap = make(map[string]bool, len(a.release.IPsToRelease))
	for _, ip := range a.release.IPsToRelease {
		excessMap[ip] = true
	}
}
n.abortNoLongerExcessIPs(excessMap)
```

**Review by gandro — COMMENTED** [review]
Good catch! Thank you.

Only some minor nits, nothing blocking.

File: pkg/ipam/node.go (hunk 4 above)
Comment (hemanthmalla) [reply pkg/ipam/node.go]: Looks much cleaner, updated this.

**Comment by hemanthmalla** [comment]
@gandro Thanks for the review 🙇  
Incorporated all the feedback, updated the commit with changes.

**Review by gandro — APPROVED** [review]

**Review by twpayne — APPROVED** [review]

**Review by christarazi — COMMENTED** [review]
Thanks for the PR. The changes LGTM, could you update the 2nd commit msg with the motivation behind the refactor?

**Comment by joestringer** [comment]
Given the current title of the PR plus the labels, if we merge & backport & release this as part of v1.11.1, then the PR will show up in the release notes under "Bugfixes" as `MaintainIPPool refactor and support for aborting release handshake`. This doesn't sound like a bugfix. Could you improve the title or add a ```` ```release-note::...``` ```` section into the PR description that more succinctly describes the bugfix? Maybe something like "Fix bug where Cilium cannot allocate unused IPs from the ENI IPAM pool"?

**Review by christarazi — APPROVED** [review]

**Comment by hemanthmalla** [comment]
@christarazi @joestringer Updated the commit message and added a release note.

**Comment by joestringer** [comment]
I see that this was previously marked `ready-to-merge` and you've only updated the commit message, so this should be good to merge now. Thanks for the fix!

**Pull request merged**